crates/core/src/network/io/reactor.rs

```diff
@@ -0,0 +1,465 @@
+                            } else {
+                                Decision::Drain(DrainReason::ConnectionDrain)
+                            }
+                            // we only drain the connection if we were the initiators of the termination
+                        },
+                        msg = incoming.next() => {
+                            self.handle_message(msg, &router).await
+                        }
+                    }
+                }
+                State::Draining {
+                    ref mut drain_timeout,
+                } => {
+                    tokio::select! {
+                        _ = drain_timeout => {
+                            debug!("Drain timed out, closing connection");
+                            Decision::Drop
+                        },
+                        msg = incoming.next() => {
+                            self.handle_message(msg, &router).await
+                        }
+                    }
+                }
+                State::WaitForEgress => {
+                    self.tx.take();
+                    if tokio::time::timeout(
+                        Duration::from_secs(5),
+                        OptionFuture::from(self.drop_egress.take())
+                            .log_slow_after(
+                                Duration::from_secs(2),
+                                tracing::Level::INFO,
+                                "Waiting for connection's egress to drain",
+                            )
+                            .with_overdue(Duration::from_secs(3), tracing::Level::WARN),
+                    )
+                    .await
+                    .is_err()
+                    {
+                        info!("Connection's egress has taken too long to drain, will drop");
+                    }
+                    Decision::Drop
+                }
+            };
+
+            match (&self.state, decision) {
+                (_, Decision::Continue) => {}
+                (State::Active, Decision::Drain(reason)) => {
+                    // send drain signal, and switch
+                    self.switch_to_draining(reason, &peer_router);
+                }
+                (State::Active, Decision::NotifyPeerShutdown) => {
+                    conn_tracker.notify_peer_shutdown(self.connection.peer());
+                    self.switch_to_draining(DrainReason::ConnectionDrain, &peer_router);
+                }
+                (_, Decision::DrainEgress) => {
+                    self.tx.take();
+                    self.state = State::WaitForEgress;
+                }
+                (State::Draining { .. }, Decision::NotifyPeerShutdown) => {
+                    conn_tracker.notify_peer_shutdown(self.connection.peer());
+                }
+                (State::Draining { .. }, Decision::Drain(reason)) => {
+                    // send drain signal, and switch
+                    self.send_drain_signal(reason);
+                }
```

Comment:
Shouldn't we have sent the drain signal already when being in state `State::Draining`?

Reply:
Correct, but we allow sending other signals while draining, for instance if there were codec errors. At the moment the peer won't act on them but it's nice to still receive then (for logging/observability purposes).